OrganizationName,OrganizationID,RepositoryURL,VulnerabilityID,PkgID,PkgName,InstalledVersion,FixedVersion,Status,Severity,CweIDs,CVSS,PrimaryURL,References,PublishedDate,LastModifiedDate,Title,Description
,,https://github.com/rustdesk/rustdesk,GHSA-c827-hfw6-qwvm,rustix@0.36.11,rustix,0.36.11,"0.35.15, 0.36.16, 0.37.25, 0.38.19",fixed,MEDIUM,,"{'ghsa': {'V3Vector': 'CVSS:3.1/AV:N/AC:L/PR:L/UI:N/S:U/C:N/I:N/A:H', 'V3Score': 6.5}}",https://github.com/advisories/GHSA-c827-hfw6-qwvm,"['https://discord.com/channels/273534239310479360/1161137828395237556', 'https://github.com/bytecodealliance/rustix', 'https://github.com/bytecodealliance/rustix/commit/31fd98ca723b93cc6101a3e29843ea5cf094e159', 'https://github.com/bytecodealliance/rustix/commit/87481a97f4364d12d5d6f30cdd025a0fc509b8ec', 'https://github.com/bytecodealliance/rustix/commit/df3c3a192cf144af0da8a57417fb4addbdc611f6', 'https://github.com/bytecodealliance/rustix/commit/eecece4a84fc58eafdc809cc2cedd374dee876a5', 'https://github.com/bytecodealliance/rustix/security/advisories/GHSA-c827-hfw6-qwvm', 'https://github.com/imsnif/bandwhich/issues/284', 'https://github.com/imsnif/bandwhich/issues/284#issuecomment-1754321993']",,,rustix's `rustix::fs::Dir` iterator with the `linux_raw` backend can cause memory explosion,"### Summary

When using `rustix::fs::Dir` using the `linux_raw` backend, it's possible for the iterator to ""get stuck"" when an IO error is encountered. Combined with a memory over-allocation issue in `rustix::fs::Dir::read_more`, this can cause quick and unbounded memory explosion (gigabytes in a few seconds if used on a hot path) and eventually lead to an OOM crash of the application.

### Details

#### Discovery

The symptoms were initially discovered in https://github.com/imsnif/bandwhich/issues/284. That post has lots of details of our investigation. See [this post](https://github.com/imsnif/bandwhich/issues/284#issuecomment-1754321993) and the [Discord thread](https://discord.com/channels/273534239310479360/1161137828395237556) for details.

#### Diagnosis

This issue is caused by the combination of two independent bugs:

1. Stuck iterator
- The `rustix::fs::Dir` iterator can fail to halt after encountering an IO error, causing the caller to be stuck in an infinite loop.
2. Memory over-allocation
- `Dir::read_more` incorrectly grows the read buffer unconditionally each time it is called, regardless of necessity.

Since `<Dir as Iterator>::next` calls `Dir::read`, which in turn calls `Dir::read_more`, this means an IO error encountered during reading a directory can lead to rapid and unbounded growth of memory use.

### PoC

```rust
fn main() -> Result<(), Box<dyn std::error::Error>> {
    // create a directory, get a FD to it, then unlink the directory but keep the FD
    std::fs::create_dir(""tmp_dir"")?;
    let dir_fd = rustix::fs::openat(
        rustix::fs::CWD,
        rustix::cstr!(""tmp_dir""),
        rustix::fs::OFlags::RDONLY | rustix::fs::OFlags::CLOEXEC,
        rustix::fs::Mode::empty(),
    )?;
    std::fs::remove_dir(""tmp_dir"")?;

    // iterator gets stuck in infinite loop and memory explodes
    rustix::fs::Dir::read_from(dir_fd)?
        // the iterator keeps returning `Some(Err(_))`, but never halts by returning `None`
        // therefore if the implementation ignores the error (or otherwise continues
        // after seeing the error instead of breaking), the loop will not halt
        .filter_map(|dirent_maybe_error| dirent_maybe_error.ok())
        .for_each(|dirent| {
            // your happy path
            println!(""{dirent:?}"");
        });

    Ok(())
}
```

### Impact

If a program tries to access a directory with its file descriptor after the file has been unlinked (or any other action that leaves the `Dir` iterator in the stuck state), and the implementation does not break after seeing an error, it can cause a memory explosion.

As an example, Linux's various virtual file systems (e.g. `/proc`, `/sys`) can contain directories that spontaneously pop in and out of existence. Attempting to iterate over them using `rustix::fs::Dir` directly or indirectly (e.g. with the `procfs` crate) can trigger this fault condition if the implementation decides to continue on errors.

An attacker knowledgeable about the implementation details of a vulnerable target can therefore try to trigger this fault condition via any one or a combination of several available APIs. If successful, the application host will quickly run out of memory, after which the application will likely be terminated by an OOM killer, leading to denial of service."
,,https://github.com/rustdesk/rustdesk,CVE-2024-32650,rustls@0.23.4,rustls,0.23.4,"0.23.5, 0.22.4, 0.21.11",fixed,HIGH,['CWE-835'],"{'ghsa': {'V3Vector': 'CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H', 'V3Score': 7.5}}",https://avd.aquasec.com/nvd/cve-2024-32650,"['https://github.com/rustls/rustls', 'https://github.com/rustls/rustls/commit/2123576840aa31043a31b0770e6572136fbe0c2d', 'https://github.com/rustls/rustls/commit/5374108df698e78c3e9ef8265cac311556be24af', 'https://github.com/rustls/rustls/commit/6e938bcfe82a9da7a2e1cbf10b928c7eca26426e', 'https://github.com/rustls/rustls/commit/ebcb4782f23b4edf9b10a7065d9e8d4362439d9c', 'https://github.com/rustls/rustls/commit/f45664fbded03d833dffd806503d3c8becd1b71e', 'https://github.com/rustls/rustls/security/advisories/GHSA-6g7w-8wpp-frhj', 'https://nvd.nist.gov/vuln/detail/CVE-2024-32650', 'https://rustsec.org/advisories/RUSTSEC-2024-0336.html']",2024-04-19T16:15:10.94Z,2024-04-19T16:19:49.043Z,Rustls is a modern TLS library written in Rust. `rustls::ConnectionCom ...,"Rustls is a modern TLS library written in Rust. `rustls::ConnectionCommon::complete_io` could fall into an infinite loop based on network input. When using a blocking rustls server, if a client send a `close_notify` message immediately after `client_hello`, the server's `complete_io` will get in an infinite loop. This vulnerability is fixed in 0.23.5, 0.22.4, and 0.21.11."
,,https://github.com/rustdesk/rustdesk,GHSA-2qv5-7mw5-j3cg,spin@0.9.6,spin,0.9.6,0.9.8,fixed,MEDIUM,,,https://github.com/advisories/GHSA-2qv5-7mw5-j3cg,"['https://github.com/mvdnes/spin-rs', 'https://github.com/mvdnes/spin-rs/issues/148', 'https://rustsec.org/advisories/RUSTSEC-2023-0031.html']",,,spin-rs initialisation failure in `Once::try_call_once` can lead to undefined behaviour for other initialisers,"`Once::try_call_once` is unsound if invoked more than once concurrently and any call fails to initialise successfully.
"
,,https://github.com/rustdesk/rustdesk,CVE-2020-26235,time@0.1.45,time,0.1.45,0.2.23,fixed,MEDIUM,['CWE-476'],"{'ghsa': {'V3Vector': 'CVSS:3.1/AV:L/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H', 'V3Score': 6.2}, 'nvd': {'V2Vector': 'AV:N/AC:M/Au:S/C:N/I:N/A:P', 'V3Vector': 'CVSS:3.1/AV:N/AC:H/PR:L/UI:N/S:U/C:N/I:N/A:H', 'V2Score': 3.5, 'V3Score': 5.3}}",https://avd.aquasec.com/nvd/cve-2020-26235,"['https://crates.io/crates/time/0.2.23', 'https://github.com/time-rs/time', 'https://github.com/time-rs/time/issues/293', 'https://github.com/time-rs/time/security/advisories/GHSA-wcg3-cvx6-7396', 'https://nvd.nist.gov/vuln/detail/CVE-2020-26235', 'https://rustsec.org/advisories/RUSTSEC-2020-0071.html']",2020-11-24T22:15:11.657Z,2020-12-03T19:42:05.373Z,Segmentation fault in time,"In Rust time crate from version 0.2.7 and before version 0.2.23, unix-like operating systems may segfault due to dereferencing a dangling pointer in specific circumstances. This requires the user to set any environment variable in a different thread than the affected functions. The affected functions are time::UtcOffset::local_offset_at, time::UtcOffset::try_local_offset_at, time::UtcOffset::current_local_offset, time::UtcOffset::try_current_local_offset, time::OffsetDateTime::now_local and time::OffsetDateTime::try_now_local. Non-Unix targets are unaffected. This includes Windows and wasm. The issue was introduced in version 0.2.7 and fixed in version 0.2.23."
